src/akushon/action/node/action_node.cpp

```diff
@@ -0,0 +1,148 @@
+#include <vector>
+
+#include "akushon/action/node/action_node.hpp"
+
+#include "akushon/action/node/action_manager.hpp"
+#include "akushon/action/model/action_name.hpp"
+#include "akushon/action/model/pose.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "tachimawari/joint/model/joint.hpp"
+#include "tachimawari_interfaces/msg/set_joints.hpp"
+#include "tachimawari_interfaces/srv/get_joints.hpp"
+
+using namespace std::chrono_literals;
+
+namespace akushon
+{
+
+ActionNode::ActionNode(
+  rclcpp::Node::SharedPtr node, std::shared_ptr<ActionManager> action_manager)
+: node(node), action_manager(action_manager), status(READY)
+{
+  set_joints_publisher = node->create_publisher<tachimawari_interfaces::msg::SetJoints>(
+    "/joint/set_joints", 10);
+
+  get_joints_client = node->create_client<tachimawari_interfaces::srv::GetJoints>(
+    "/joint/get_joints");
+}
+
+bool ActionNode::is_action_exist(int action_id) const
+{
+  if (action_manager->get_action(action_id).get_name() != "") {
+    return true;
+  }
+
+  return false;
+}
+
+bool ActionNode::is_action_exist(std::string action_name) const
+{
+  return is_action_exist(ActionName::map.at(action_name));
+}
+
+int ActionNode::get_status() const
+{
+  return status;
+}
+
+bool ActionNode::start(std::string action_name)
+{
+  return start(ActionName::map.at(action_name));
+}
+
+bool ActionNode::start(int action_id)
+{
+  while (!get_joints_client->wait_for_service(1s)) {
+    if (rclcpp::ok()) {
+      // service not available, waiting again...
+    } else {
+      // Interrupted while waiting for the service. Exiting.
+      return false;
+    }
+  }
+
+  auto result = get_joints_client->async_send_request(
+    std::make_shared<tachimawari_interfaces::srv::GetJoints::Request>());
+
+  if (rclcpp::spin_until_future_complete(node, result) ==
+    rclcpp::FutureReturnCode::SUCCESS)
+  {
+    Pose pose("initial_pose");
+    std::vector<tachimawari::joint::Joint> joints;
+
+    for (const auto & joint : result.get()->joints) {
+      joints.push_back(
+        tachimawari::joint::Joint(joint.id, joint.position));
+    }
+    pose.set_joints(joints);
+
+    action_manager->start(action_id, pose);
+    status = PLAYING;
+  } else {
+    // Failed to call service
+    return false;
+  }
+
+  return true;
+}
+
+void ActionNode::process(int time)
+{
+  action_manager->process(time);
+
+  if (action_manager->is_playing()) {
+    publish_joints();
+  } else {
+    status = READY;
+  }
+}
+
+std::string ActionNode::get_node_prefix() const
+{
+  return "action";
+}
+
+void ActionNode::publish_joints()
+{
+  std::vector<tachimawari_interfaces::msg::Joint> joints;
+
+  for (const auto & joint : action_manager->get_joints()) {
+    auto joint_msg = tachimawari_interfaces::msg::Joint();
+
+    joint_msg.id = joint.get_id();
+    joint_msg.position = joint.get_position();
+
+    joints.push_back(joint_msg);
+  }
+
+  auto joints_msg = tachimawari_interfaces::msg::SetJoints();
+  joints_msg.joints = joints;
```

Comment:
if performance is important, we could do this instead
```c++
auto joints_msg = tachimawari_interfaces::msg::SetJoints();

const auto & joints = action_manager->get_joints();
auto & joint_msgs = joints_msg.joints;

joint_msgs.resize(joints.size());
for (size_t i = 0; i < joints.size() && i < joints_msg.size(); ++i) {
  joint_msgs[i].id = joints[i].get_id();
  // ...
}
```

Reply:
done